### PR TITLE
Storage pubsub refactor

### PR DESCRIPTION
**Pull request by akshaisarma**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by akshaisarma** [comment]
Adds new Storage abstractions. Breaks PubSubMessage and Subscriber interfaces:

1. There is no longer a sequence field
2. The `content` field is now a `byte[]` (from `String`)
3. Two new Metadata.Signal enums - `REPLAY` and `CUSTOM`
4. Subscriber#commit and Subscriber#fail no longer take a sequence
5. PubSubMessage's hashCode is different as is its toJSON and fromJSON since sequence is removed and content is a `byte[]`

File: src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java [hunk 1]
```diff
@@ -39,7 +37,7 @@
     /**
      * A Map of messages that have not been committed so far.
      */
-    protected Map<Pair<String, Integer>, PubSubMessage> uncommittedMessages = new HashMap<>();
+    protected Map<String, PubSubMessage> uncommittedMessages = new HashMap<>();
 
     /**
      * Creates an instance of this class with the given max for commited messages.
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java]: committed

File: src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java [hunk 2]
```diff
@@ -72,9 +70,8 @@ public PubSubMessage receive() throws PubSubException {
      * nulls till some messages are commited.
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/pubsub/BufferingSubscriber.java]: committed

File: src/main/java/com/yahoo/bullet/storage/StorageManager.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    public abstract CompletableFuture<Boolean> putAll(Map<String, byte[]> data);
+
+    /**
+     * Retrieves all the IDs stored with {@link #putString(String, String)} or {@link #put(String, byte[])} from the
+     * storage as byte[].
+     *
+     * @return A {@link CompletableFuture} that resolves to a {@link Map} of IDs to their stored values as byte[] or
+     *         null if no data is present. It completes exceptionally if there were issues.
+     */
+    public abstract CompletableFuture<Map<String, byte[]>> getAll();
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/storage/StorageManager.java]: Why return null when data isn't present?

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/storage/StorageManager.java]: You want it to resolve to an empty map instead?

Comment (0aix) [reply src/main/java/com/yahoo/bullet/storage/StorageManager.java]: Yea, that seems more natural in my opinion

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/storage/StorageManager.java]: I didn't want to because get resolves to null if there's no data

**Pull request merged**